### PR TITLE
util/winutil: simplify dropping privileges after use 

### PR DESCRIPTION
**Pull request by awly**
To safely request and drop privileges, runtime.Lock/UnlockOSThread and
windows.Impersonate/RevertToSelf should be called. Add these calls to
winutil.EnableCurrentThreadPrivilege so that callers don't need to worry
about it.

Updates https://github.com/tailscale/corp/issues/15488

### CODE REVIEW

File: util/winutil/winutil_windows.go [hunk 1]
```diff
@@ -225,9 +225,20 @@ func isSIDValidPrincipal(uid string) bool {
...
+	return adjustCurrentThreadPrivilege(name, windows.SE_PRIVILEGE_ENABLED)
+}
+
+// DisableCurrentThreadPrivilege disabled the named privilege in the current
+// thread access token.
+func DisableCurrentThreadPrivilege(name string) error {
+	return adjustCurrentThreadPrivilege(name, windows.SE_PRIVILEGE_REMOVED)
+}
+
+func adjustCurrentThreadPrivilege(name string, attr uint32) error {
```
Comment (bradfitz) [inline util/winutil/winutil_windows.go:241]: It doesn't really matter for our current sole user, but if this ever becomes something we do more hotly, having Enable return a `disable func()` would let us do the UTF16PtrFromString and LookupPrivilegeValue just once.

Also, then the Enable can do its own runtime.LockOSThread() (they nest nowadays!) that the disable/release func could undo, removing that potential footgun for callers who havent' already done the LockOSThread.

If not, at least document callers must LockOSThread on their own? (it looks like?)

**Review by dblohm7 — CHANGES_REQUESTED** [review]
There is a much simpler and much more Win32 idiomatic way to do this (and you'll notice that we already do so in many places): Call [ImpersonateSelf](https://learn.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-impersonateself) before enabling the privileges, and then immediately doing a `defer` to `RevertToSelf`.

The former API is explicitly designed such that it allows you to set a bunch of privileges on the _impersonated_ token, instead of the primary token, and then the reversion cancels all of that in one fell swoop.

Of course, you also need to ensure that the OS thread has been locked, but that already need to be done when enabling privileges anyway.

File: util/winutil/winutil_windows.go (hunk 1 above)
Comment (awly) [reply util/winutil/winutil_windows.go]: Moved ImpersonateSelf/LockOSThread into the function.
Per @dblohm7's comment, we can use `RevertToSelf` to drop the privilege, no need to explicitly disable it.

**Comment by awly** [comment]
Thanks for the explanation @dblohm7, TIL.
Updated the PR to move ImpersonateSelf/LockOSThread into EnableCurrentThreadPrivilege to make it less error-prone in the future.
@bradfitz PTAL

File: util/winutil/winutil_windows.go [hunk 2]
```diff
@@ -225,47 +225,56 @@ func isSIDValidPrincipal(uid string) bool {
...
+// done with the privileged task.
+func EnableCurrentThreadPrivilege(name string) (disable func() error, err error) {
+	runtime.LockOSThread()
+
+	if err := windows.ImpersonateSelf(windows.SecurityImpersonation); err != nil {
+		runtime.UnlockOSThread()
+		return nil, err
+	}
+	disable = func() error {
+		runtime.UnlockOSThread()
```
Comment (bradfitz) [inline util/winutil/winutil_windows.go:240]: I think you need to defer this so the RevertToCall runs first, no?

File: util/winutil/winutil_windows.go [hunk 3]
```diff
@@ -225,47 +225,56 @@ func isSIDValidPrincipal(uid string) bool {
 	}
 }
 
-// EnableCurrentThreadPrivilege enables the named privilege
-// in the current thread access token.
-func EnableCurrentThreadPrivilege(name string) error {
+// EnableCurrentThreadPrivilege enables the named privilege in the current
+// thread access token. The current goroutine is also locked to the OS thread
+// (runtime.LockOSThread). Callers should the returned disable function when
```
Comment (bradfitz) [inline util/winutil/winutil_windows.go:230]: not an RFC but s/should/must/?

**Review by bradfitz — APPROVED** [review]

**Pull request merged**